libwacom/libwacom-database.c

```diff
@@ -279,8 +323,88 @@ parse_stylus_id(const char *str, WacomStylusId *id)
 	return ret;
 }
 
+static char *
+string_or_fallback (GKeyFile *keyfile, const char *group,
+		    const char *key, const char *fallback) {
+	g_autoptr(GError) local_error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &local_error))
+		return g_key_file_get_string(keyfile, group, key, NULL);
+
+	return g_strdup(fallback);
+}
+
+static int
+int_or_fallback (GKeyFile *keyfile, const char *group,
+		const char *key, int fallback) {
+	g_autoptr(GError) local_error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &local_error))
+		return g_key_file_get_integer(keyfile, group, key, NULL);
+
+	return fallback;
+}
+
+static bool
+boolean_or_fallback (GKeyFile *keyfile, const char *group,
+		 const char *key, bool fallback, WacomStylus *stylus) {
+	g_autoptr(GError) error = NULL;
+	if (g_key_file_has_key(keyfile, group, key, &error)) {
+		return g_key_file_get_boolean(keyfile, group, key, NULL);
+	}
+	if (error && error->code == G_KEY_FILE_ERROR_INVALID_VALUE)
+		g_warning ("Stylus %s (%s) %s\n", stylus->name, group, error->message);
+
+	return fallback;
+}
+
+gchar **
+garray_to_string_list(GArray *array)
+{
+	gchar **values;
+	WacomStylusId *id;
+	char buffer[33];
+	bool ret;
+
+	if (array == NULL)
+		return NULL;
+
+	values = g_new (gchar *, array->len + 1);
+	for (guint i = 0; i < array->len; i++) {
+		id = &g_array_index(array, WacomStylusId, i);
+		ret = safe_itoa_base(id->tool_id, buffer, sizeof(buffer), 16);
+		values[i] = strdup(buffer);
```

Comment:
`safe_itoa_base` always returns true so the `ret` assignment is pointless (and we don't check it here anyway). But since this is appears to be the only place we need it we can skip that helper altogether and make it a:

```suggestion
		values[i] = g_strdup_printf("0x%08x", id->tool_id);
```